File: cbta97net-fix-editar-alumno/cbta97net-front/src/app/nucleo/servicios/estructura/aula.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from 'environments/environment';
import { Aula } from '@modelos/comunes/aula.model';

@Injectable({
    providedIn: 'root'
})
export class AulaServicio {

    private readonly URL_AULAS = `${environment.apiUrl}/aulas`;

    /**
     * Constructor por defecto. Permite asignar clientes HTTP mockeables para
     * realizacion de pruebas.
     * @param httpClient cliente para realizar peticiones HTTP.
     */
    constructor(private readonly http: HttpClient) { }

    /**
     * Registra una nueva aula en el sistema.
     * Requiere rol: AULAS_CREATE
     * @param aula Objeto con la clave y datos del aula.
     * @returns Un Observable con el aula creada (incluyendo su ID).
     */
    registrarAula(aula: Aula): Observable<Aula> {
        return this.http.post<Aula>(this.URL_AULAS, aula);
    }

    /**
     * Elimina un aula mediante su identificador numérico.
     * Requiere rol: AULAS_DELETE
     * @param id ID del aula a eliminar.
     * @returns Un Observable con el mensaje de confirmación.
     */
    eliminarAulaPorId(id: number): Observable<{ message: string }> {
        return this.http.delete<{ message: string }>(`${this.URL_AULAS}/${id}`);
    }

    /**
     * Elimina un aula utilizando su clave única.
     * Requiere rol: AULAS_DELETE
     * @param clave Clave alfanumérica del aula.
     * @returns Un Observable con el mensaje de confirmación.
     */
    eliminarAulaPorClave(clave: string): Observable<{ message: string }> {
        return this.http.delete<{ message: string }>(`${this.URL_AULAS}/clave/${clave}`);
    }

    /**
     * Recupera el listado completo de aulas registradas.
     * @returns Un Observable con un arreglo de objetos Aula.
     */
    obtenerTodasLasAulas(): Observable<Aula[]> {
        return this.http.get<Aula[]>(this.URL_AULAS);
    }
}

File: cbta97net-fix-editar-alumno/cbta97net-backend/src/main/java/mx/edu/cbta/sistemaescolar/estructura/service/AulaService.java
package mx.edu.cbta.sistemaescolar.estructura.service;

import mx.edu.cbta.sistemaescolar.estructura.domain.exception.AulaNoEncontradaException;
import mx.edu.cbta.sistemaescolar.estructura.domain.exception.AulaDuplicadaException;
import mx.edu.cbta.sistemaescolar.estructura.domain.exception.RegistrarAulaException;
import mx.edu.cbta.sistemaescolar.estructura.domain.exception.EliminarAulaException;
import mx.edu.cbta.sistemaescolar.estructura.dto.AulaDTO;

import java.util.List;

/**
 * Interfaz de servicio que define las operaciones de negocio para la gestión de Aulas.
 * Proporciona métodos para el mantenimiento (CRUD) y consulta de espacios físicos
 * del plantel académico.
 */
public interface AulaService {

    /**
     * Busca un aula específica mediante su identificador único numérico.
     * * @param idAula El identificador único de la base de datos.
     * @return {@link AulaDTO} con los datos del aula encontrada.
     * @throws AulaNoEncontradaException Si no existe un registro con el ID proporcionado.
     */
    AulaDTO obtenerAulaPorId(Long idAula) throws AulaNoEncontradaException;

    /**
     * Recupera la información de un aula utilizando su clave alfanumérica única.
     * * @param clave La clave identificadora del aula (ej. "A-101").
     * @return {@link AulaDTO} con los datos correspondientes.
     * @throws AulaNoEncontradaException Si la clave no coincide con ningún registro.
     */
    AulaDTO obtenerAulaPorClave(String clave) throws AulaNoEncontradaException;

    /**
     * Registra una nueva aula en el sistema.
     * Realiza validaciones de formato (sin espacios en la clave) y unicidad.
     * * @param aula Objeto DTO con la información de la nueva aula.
     * @return {@link AulaDTO} El registro persistido con su ID generado.
     * @throws RegistrarAulaException Si los datos tienen un formato inválido.
     * @throws AulaDuplicadaException Si la clave o el ID ya existen en el sistema.
     */
    AulaDTO registrarAula(AulaDTO aula) throws RegistrarAulaException, AulaDuplicadaException;

    /**
     * Recupera el catálogo completo de aulas registradas en el plantel.
     * * @return List de {@link AulaDTO}, o una lista vacía si no hay registros.
     */
    List<AulaDTO> obtenerTodasLasAulas();

    /**
     * Elimina de forma lógica o física un aula mediante su ID.
     * * @param aulaId El identificador del aula a remover.
     * @throws AulaNoEncontradaException Si el ID no existe.
     * @throws EliminarAulaException Si el aula tiene dependencias (horarios/grupos) que impiden su borrado.
     */
    void eliminarAulaPorId(Long aulaId) throws AulaNoEncontradaException, EliminarAulaException;

    /**
     * Elimina un aula utilizando su clave única.
     * Requiere que el repositorio soporte operaciones transaccionales para campos no-ID.
     * * @param clave La clave del aula a eliminar.
     * @throws AulaNoEncontradaException Si la clave no está registrada.
     * @throws EliminarAulaException Si ocurre un error de integridad referencial.
     */
    void eliminarAulaPorClave(String clave) throws AulaNoEncontradaException, EliminarAulaException;
}
